### PR TITLE
CONSVC-1901 Add async apis to remote settings client

### PR DESCRIPTION
**Pull request by quiiver**
* Moved provider initialization out of the class __init__ method and
  into an abstract class method called `initialize`
* Added `httpx` as an async `requests` alternative

### CODE REVIEW

File: merino/providers/__init__.py [hunk 1]
```diff
@@ -1,21 +1,33 @@
...
+import logging
+from timeit import default_timer as timer
+
 from merino.providers.adm import Provider as AdmProvider
 from merino.providers.base import BaseProvider
 
 providers: dict[str, BaseProvider] = {}
 default_providers: list[BaseProvider] = []
 
+logger = logging.getLogger("merino.providers")
```
Comment (ncloudioj) [inline merino/providers/__init__.py:11]: nit: you could do `logging.getLogger(__name__)` as well.

File: merino/providers/__init__.py [hunk 2]
```diff
@@ -1,21 +1,33 @@
...
+    await asyncio.gather(*[p.initialize() for p in providers.values()])
     default_providers.extend([p for p in providers.values() if p.enabled_by_default()])
+    logger.info(
+        "Provider initialization complete",
+        extra={"providers": [*providers.keys()], "elapsed": timer() - start},
+    )
 
 
-def get_providers() -> tuple[dict[str, BaseProvider], list[BaseProvider]]:
+async def get_providers() -> tuple[dict[str, BaseProvider], list[BaseProvider]]:
```
Comment (ncloudioj) [inline merino/providers/__init__.py:30]: Hmm, why is this an async func?

File: merino/providers/adm.py [hunk 3]
```diff
@@ -31,23 +32,36 @@ class Provider(BaseProvider):
...
-        suggest_settings = rs.get("main", "quicksuggest")
+        suggest_settings = await rs.get("main", "quicksuggest")
         data_items = [i for i in suggest_settings if i["type"] == "data"]
-        for item in data_items:
-            res = rs.fetch_attachment(item["attachment"]["location"])
+
+        fetch_tasks = [
+            rs.fetch_attachment(item["attachment"]["location"]) for item in data_items
+        ]
+        for routine in as_completed(fetch_tasks):
```
Comment (ncloudioj) [inline merino/providers/adm.py:47]: nit: can you rename `routine` to `done_task` please?

File: merino/providers/__init__.py (hunk 1 above)
Comment (quiiver) [reply merino/providers/__init__.py:11]: sure

File: merino/providers/base.py [hunk 4]
```diff
@@ -7,6 +7,15 @@ class BaseProvider(ABC):
     Abstract class for suggestion providers.
     """
 
+    @abstractmethod
+    async def initialize(self) -> bool:
```
Comment (ncloudioj) [inline merino/providers/base.py:11]: Do we need a return value for this method? Looks like the provider in `adm.py` doesn't return a boolean. When getting unrecoverable initialization errors, normally the application would log them and exit. Unless we want to handle those init errors, I don't think the return value would be necessary here.

Comment (quiiver) [reply merino/providers/base.py:11]: oh yeah i meant to drop that. thanks

**Review by ncloudioj — CHANGES_REQUESTED** [review]
This looks great. Having a few nits.

I've tested it locally, the async concurrent loading cuts the loading time from 3.5s to 0.7s on my laptop!!!

File: merino/providers/__init__.py (hunk 2 above)
Comment (ncloudioj) [reply merino/providers/__init__.py]: Ah, i see. Let's drop it. The test/util function can be a reg function as well, we don't need to call `initialize()` for those.

**Review by ncloudioj — APPROVED** [review]
Good stuff! 🚢

**Comment by quiiver** [comment]
this fixes https://mozilla-hub.atlassian.net/browse/CONSVC-1901

**Pull request merged**